perma_web/perma/tasks.py

```diff
@@ -266,32 +270,56 @@ def scroll_browser(browser):
         # URLError: the headless browser has gone away for some reason.
         pass
 
-def run_in_frames(link, browser, func, output_collector=None):
+def get_page_source(browser):
+    """ 
+        Get page source.
+        Use JS rather than browser.page_source so we get the parsed, properly formatted DOM instead of raw user HTML.
+    """
+    return browser.execute_script("return document.documentElement.outerHTML")
+
+def parse_page_source(source):
+    """
+        Return page source as a parsed PyQuery object for querying.
+         
+        PyQuery here works the same as `$(source)` in jQuery. So for example you can do `parsed_source(selector)` 
+        with the returned value to get a list of LXML elements matching the selector.
+    """
+    return PyQuery(source, parser='html')
+
+def get_dom_tree(browser):
+    with browser_running(browser):
+        return parse_page_source(get_page_source(browser))
```

Comment:
Thanks for checking that! Good to know it's not long relative to our total capture time.

As background I guess I mean it's slow on some sort of order-of-magnitude programmer intuition scale. Like if we rank things by order of magnitude runtime ...

typical python function < Redis access < typical database query ~ disk access < S3 access ~ typical remote API access

... then you have to fret exponentially more about how frequently each of those is used in a given algorithm as you move to the right -- pretty much no loops allowed after the second `<`, or it creates a human-noticeable delay. `get_dom_tree` is all the way on the right on this scale -- but that's a win, because it replaces multiple functions that were all the way on the right, and now those are over on the left at "typical python function."